xeth/xeth.go

```diff
@@ -304,25 +327,61 @@ func (self *XEth) Transact(toStr, valueStr, gasStr, gasPriceStr, codeStr string)
 		tx = types.NewTransactionMessage(to, value.BigInt(), gas.BigInt(), price.BigInt(), data)
 	}
 
-	var err error
-	state := self.eth.ChainManager().TxState()
-	if balance := state.GetBalance(key.Address()); balance.Cmp(tx.Value()) < 0 {
-		return "", fmt.Errorf("insufficient balance. balance=%v tx=%v", balance, tx.Value())
-	}
-	nonce := state.GetNonce(key.Address())
-
+	state := self.chainManager.TransState()
+	nonce := state.GetNonce(from)
 	tx.SetNonce(nonce)
-	tx.Sign(key.PrivateKey)
 
-	err = self.eth.TxPool().Add(tx)
-	if err != nil {
+	if err := self.sign(tx, from, false); err != nil {
+		return "", err
+	}
+	if err := self.eth.TxPool().Add(tx); err != nil {
 		return "", err
 	}
-	state.SetNonce(key.Address(), nonce+1)
+	state.SetNonce(from, nonce+1)
+
+	if contractCreation {
+		addr := core.AddressFromMessage(tx)
+		pipelogger.Infof("Contract addr %x\n", addr)
+	}
 
 	if types.IsContractAddr(to) {
 		return toHex(core.AddressFromMessage(tx)), nil
 	}
-
 	return toHex(tx.Hash()), nil
 }
+
+func (self *XEth) sign(tx *types.Transaction, from []byte, didUnlock bool) error {
+	sig, err := self.accountManager.Sign(accounts.Account{Address: from}, tx.Hash())
+	if err == accounts.ErrLocked {
+		if didUnlock {
+			return fmt.Errorf("sender account still locked after successful unlock")
```

Comment:
Why can this happen? Race? If it's an expected race then the error message should be a bit different, e.g. "account unlocked, please unlock again".